Add post update hook

ermanno opened the pull request:
### Notes
Add a hook to call the post-update script.

### Tests
- Created a cluster with a post update script, using the CLI code in https://github.com/aws/aws-parallelcluster/pull/4501.
- Logged into the head node with `pcluster ssh` and applied this patch's changes.
- Triggered a successful update, verifying that the post update script has been executed.
- Verified that for other states (such as `UPDATE_COMPLETE`) the script correctly exits with zero return code.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ermanno committed:
Add postupdate hook

enrico-usai commented on cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:
```diff
@@ -33,6 +37,13 @@ function download_run (){
 
+function get_stack_status () {
+  region=$(grep cfn_region /etc/parallelcluster/cfnconfig | sed 's/cfn_region=//' | xargs) || return 1
+  stack_name=$(grep stack_name /etc/parallelcluster/cfnconfig | sed 's/stack_name=//' | xargs) || return 1
```
you can compact grep and sed in the same command with:
`sed -n 's/stack_name=//p' | xargs`

enrico-usai reviewed (commented):
Memo: add changelog entry e.g.

> Add possibility to execute a custom script in the head node during the update of the cluster.

enrico-usai commented on cookbooks/aws-parallelcluster-scheduler-plugin/recipes/update_head_node.rb:
```diff
@@ -25,3 +25,9 @@
 execute_event_handler 'HeadClusterUpdate' do
   event_command(lazy { node['cluster']['config'].dig(:Scheduling, :SchedulerSettings, :SchedulerDefinition, :Events, :HeadClusterUpdate, :ExecuteCommand, :Command) })
 end
+
```
It's useful to add a comment explaining that the update cfnconfig will be used by post update custom scripts

enrico-usai commented on cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:
```diff
@@ -33,6 +39,13 @@ function download_run (){
+function get_stack_status () {
+  region=$(sed -n 's/cfn_region=//p' /etc/parallelcluster/cfnconfig | xargs) || return 1
+  stack_name=$(sed -n 's/stack_name=//p' /etc/parallelcluster/cfnconfig | xargs) || return 1
+  stack_status=$(aws cloudformation describe-stacks --region "${region}" --stack-name "${stack_name}" --query Stacks[0].StackStatus | xargs) || return 1
```
an alternative to `xargs` in this case is `--output text`

enrico-usai commented on cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:
```diff
@@ -33,6 +39,13 @@ function download_run (){
+  region=$(sed -n 's/cfn_region=//p' /etc/parallelcluster/cfnconfig | xargs) || return 1
+  stack_name=$(sed -n 's/stack_name=//p' /etc/parallelcluster/cfnconfig | xargs) || return 1
+  stack_status=$(aws cloudformation describe-stacks --region "${region}" --stack-name "${stack_name}" --query Stacks[0].StackStatus | xargs) || return 1
+  echo $stack_status
```
let's add brackets `${stack_status}`

enrico-usai commented on cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:
```diff
@@ -56,6 +69,24 @@ function run_postinstall () {
+  stack_status=$(get_stack_status) || error_exit "Failed to get the stack status"
+
+  if [ "${stack_status}" != "UPDATE_IN_PROGRESS" ]; then
+    echo "postupdate called with CFN state: ${stack_status}, doing nothing"
```
`Post update hook called with CFN stack in state ..., doing nothing`

enrico-usai commented on cookbooks/aws-parallelcluster-config/templates/default/init/fetch_and_run.erb:
```diff
@@ -56,6 +69,24 @@ function run_postinstall () {
+    else
+      download_run ${cfn_postupdate}
+    fi
+  fi || error_exit "Failed to run postupdate"
```
`Failed to run post update hook`

ermanno replied:
I have no strong feelings on this, I used this phrasing to align on what we do for the other custom actions, as we use the phrasing `Failed to run preinstall` and `Failed to run postinstall`.

ermanno replied:
Actually, I'll make this change here and then change the other two messages to match, in a separate PR, as `post update hook` is clearer.

enrico-usai commented on cookbooks/aws-parallelcluster-slurm/recipes/update_head_node.rb:
```diff
@@ -205,3 +205,9 @@ def update_nodes_in_queue(strategy, queues)
   command "#{node['cluster']['cookbook_virtualenv_path']}/bin/supervisorctl start clustermgtd"
   not_if { ::File.exist?(node['cluster']['previous_cluster_config_path']) && !are_queues_updated? }
 end
+
```
same here

ermanno committed:
Small cleanups in fetch_and_run template

Some small cleanups:
- Delete unused RC variable
- Fix indentation
- Rename messages for pre/post install hook failures
